---
title: "Glioma biomarkers to distinguish high-risk from low-risk patients"
author: "Beatriz N. Leitão, André Veríssimo, Alexandra M. Carvalho and Susana Vinga"
date: "March, 2025"
output:
  html_document: 
    number_sections: yes
    toc: yes
editor_options:
  chunk_output_type: inline
---

```{r init, include=FALSE}

# Trick to store functions in search path, but not enviromnent
local({
  env <- new.env()
  env$fun_source <- function(file_name) {
    env_name <- sprintf("my_functions:%s", gsub(" ", "_", basename(file_name)))
    if (env_name %in% search()) {
      detach(env_name, character.only = TRUE, unload = TRUE)
    }
    sys.source(file_name, envir = attach(NULL, name = env_name))
  }
  lapply(search()[search() %in% "fun_source"], detach, character.only = TRUE, unload = TRUE)
  attach(env, name = "fun_source")
})

# Run garbage colector after each chunk
knitr::knit_hooks$set(
  after_chunk = function(options) {
    invisible(gc())
  }
)
```

# Install packages

```{r Install_packages, results = 'hide', message=FALSE, warning=FALSE}
#renv::restore()
```

# Load libraries

```{r Load_libraries, results=FALSE, message=FALSE, warning=FALSE}
#library(TCGAbiolinks)
library(dplyr)
library(tibble)
# library(tidyverse)
library(DT)
library(SummarizedExperiment)
library(survival)
library(survminer)
# library(readr)
library(glmnet)
library(caret)
library(openxlsx)
library(biomaRt)
library(writexl)
library(edgeR)
library(pROC)
library(caTools)
library(survivalROC)
library(risksetROC)
library(lattice)
library(survMisc)
library(SIS)
library(VennDiagram)
# library(glmSparseNet)
library(randomForestSRC)
library(grf)
library(Boruta)
library(timeROC)
library(qvalue)
library(clusterProfiler)
library(org.Hs.eg.db)
library(enrichplot)
library(pathview)
library(UpSetR)
library(rsample)
library(pec)
```

# Prepare the environment, report chunk status and sessioninfo

```{r Prepare_the_environment_report_chunk_status_and_sessioninfo}

# clean current environment
rm(list = ls())

# Capture the sessionInfo() output and write it to a file
writeLines(capture.output(sessionInfo()), con = "sessioninfo.txt")

fun_source("Functions/Logger.R")

# write on the chunk end time file
log_chunk_end_time("Prepare the environment, report chunk status and sessioninfo", log_file = "chunk_times_log.txt")
```

# Download functions

```{r Download_functions}
# download the functions
fun_source("Functions/Data_preparation.R")
fun_source("Functions/Model_evaluation.R")
fun_source("Functions/Model_fitting.R")

# write on the chunk end time file
log_chunk_end_time("Download functions", log_file = "chunk_times_log.txt")
```

# Folder to store results

```{r Folder_to_store_results}

# Ensure the 'Results' directory exists
if (!dir.exists("CGGA_Results")) {
  dir.create("CGGA_Results")
}

# write on the chunk end time file
log_chunk_end_time("Folder to store results", log_file = "chunk_times_log.txt")
```

# Download data CGGA

```{r Download_data, results = 'hide', message=FALSE, warning=FALSE}
# Cache to use in memoise
cd <- cachem::cache_disk(here::here(".", "memoise-cache"), max_size = 3 * 1024^3)

# download_data <- memoise::memoise(
#   function() {
#     query_RNA <- TCGAbiolinks::GDCquery(
#       project = c("TCGA-GBM", "TCGA-LGG"),
#       data.category = "Transcriptome Profiling",
#       data.type = "Gene Expression Quantification",
#       workflow.type = "STAR - Counts",
#       experimental.strategy = ("RNA-Seq")
#     )
#     # Check if files exist before downloading
#     if (!dir.exists("GDCdata")) {
#       TCGAbiolinks::GDCdownload(query = query_RNA)
#     }
# 
#     TCGAbiolinks::GDCprepare(query = query_RNA)
#   },
#   cache = cd
# )
# 
# TCGA_RNA <- download_data()
# saveRDS(TCGA_RNA, file = "TCGA_RNA.rds")

# library(googledrive)

# Ensure that we don't use authentication (for public files)
#drive_deauth()

# Now download the file directly
#drive_download(as_id("1UkGI-yjHDu_HQFL4dmwW-T9vde2JnD3J"), 
#               path = "TCGA_RNA.rds", 
#               overwrite = TRUE)

#TCGA_RNA <- readRDS("TCGA_RNA.rds")

# query_RNA <- TCGAbiolinks::GDCquery(
#   project = c("TCGA-GBM", "TCGA-LGG"),
#   data.category = "Transcriptome Profiling",
#   data.type = "Gene Expression Quantification",
#   workflow.type = "STAR - Counts",
#   experimental.strategy = "RNA-Seq"
# )
# 
# if (!dir.exists("GDCdata")) {  
#   TCGAbiolinks::GDCdownload(query = query_RNA)  
# }
# 
# download_data <- memoise::memoise(
#   function() {
#     TCGAbiolinks::GDCprepare(query = query_RNA)  # ✅ This is the only step cached
#   },
#   cache = cd  # ✅ Persistent caching
# )
# 
# TCGA_RNA <- download_data()

####################################################################################################
if (!dir.exists("CGGA_DATA")) {
  # Define URLs for clinical and expression data
  clinical_url <- "https://www.cgga.org.cn/download?file=download/20200506/CGGA.mRNAseq_693_clinical.20200506.txt.zip&type=mRNAseq_693_clinical&time=20200506"
  expression_url <- "https://www.cgga.org.cn/download?file=download/20200506/CGGA.mRNAseq_693.RSEM-genes.20200506.txt.zip&type=mRNAseq_693&time=20200506"

  # Create the data directory if it doesn't exist
  dir.create("CGGA_DATA", showWarnings = FALSE)

  # Download clinical data
  download.file(clinical_url, destfile = "CGGA_DATA/CGGA_clinical_693.zip", mode = "wb")

  # Download expression data
  download.file(expression_url, destfile = "CGGA_DATA/CGGA_expression_693.zip", mode = "wb")

  # Unzip both
  unzip("CGGA_DATA/CGGA_clinical_693.zip", exdir = "CGGA_DATA")
  unzip("CGGA_DATA/CGGA_expression_693.zip", exdir = "CGGA_DATA")
  
  message("Download and extraction completed.")
} else {
  message("Folder 'CGGA_DATA' already exists. Skipping download.")
}

# Load the clinical data
clinical_df <- read.delim("CGGA_DATA/CGGA.mRNAseq_693_clinical.20200506.txt", 
                          header = TRUE, 
                          sep = "\t", 
                          stringsAsFactors = FALSE)

# Load the expression data
expression_df <- read.delim("CGGA_DATA/CGGA.mRNAseq_693.RSEM-genes.20200506.txt", 
                            header = TRUE, 
                            sep = "\t", 
                            stringsAsFactors = FALSE)

# Set gene names as rownames
rownames(expression_df) <- expression_df[, 1]
expression_df <- expression_df[, -1]

# Transpose
expression_df <- as.data.frame(t(expression_df))

# Add patient ID column
expression_df$patient <- rownames(expression_df)
rownames(expression_df) <- NULL

# Reorder columns: put CGGA_ID first, then the rest
expression_df <- expression_df[, c("patient", setdiff(names(expression_df), "patient"))]


# write on the chunk end time file
log_chunk_end_time("Download data", log_file = "chunk_times_log.txt")
```


# Prepare survival data

```{r Prepare_survival_data}
# # Extract survival data of interest
# columns_to_extract <- c("patient", "project_id", "vital_status", "days_to_death", "days_to_last_follow_up")
# survival_data <- createSurvivalDataFrame(columns_to_extract, TCGA_RNA)
# 
# 
# # Clean and prepare survival data
# survival_data <- cleanSurvivalData(survival_data)

###################################################################################################

# Extract survival data of interest and rename desired columns
survival_data <- clinical_df[, c("CGGA_ID", "Histology", "OS", "Censor..alive.0..dead.1.", "IDH_mutation_status")]
colnames(survival_data) <- c("patient", "project_id", "days", "vital_status", "IDH_mutation_status")

# Remove rows with any NA values
survival_data <- na.omit(survival_data)

# write on the chunk end time file
log_chunk_end_time("Prepare survival data", log_file = "chunk_times_log.txt")
```

<!-- # Assign a new classification to the data -->

<!-- ```{r Assign_new_classification_to_the_data, results = 'hide', message=FALSE, warning=FALSE} -->
<!-- # new classification --> -->
<!-- new_classification <- readr::read_csv("SIMPLIFIED_CLASSIFICATION_TCGA_2016_2021.csv") #download the classification datafram -->
<!-- names(new_classification)[names(new_classification) == "Patient_ID"] <- "patient" # remane the Patient_ID column -->
<!-- new_classification <- subset(new_classification, select = -c(TCGA.histological.type, classification.2016)) # delete unnecessary columns -->

<!-- # Merge the dataframes to add the disease information keeping only common patients -->
<!-- survival_data <- merge(survival_data, new_classification, by = "patient") -->
<!-- survival_data <- subset(survival_data, select = -project_id) # remove old classification -->
<!-- names(survival_data)[names(survival_data) == "classification.2021"] <- "project_id" # aplly the old name to the new classification -->


<!-- # write on the chunk end time file -->
<!-- log_chunk_end_time("Assign a new classification to the data", log_file = "chunk_times_log.txt") -->
<!-- ``` -->

# Select and prepare expression data

```{r Select_and_prepare_expression_data}
# # Extrair dataframe com a expressão dos genes de interesse
# genes_expression <- createGeneExpressionDataframe("protein_coding", "unstranded", TCGA_RNA, survival_data)
# #genes_expression <- createGeneExpressionDataframe("lncRNA", "unstranded", TCGA_RNA, survival_data)
# #genes_expression <- createGeneExpressionDataframe("miRNA", "unstranded", TCGA_RNA, survival_data)       #<------------ indicar os dados 
# 
# # Clean the gene expression dataframe
# genes_expression <- cleanGeneExpressionData(genes_expression)

#######################################################
# Clean the gene expression dataframe
genes_expression <- cleanGeneExpressionData(expression_df)

# write on the chunk end time file
log_chunk_end_time("Select and prepare expression data", log_file = "chunk_times_log.txt")
```

```{r clean_up_tgca_rna, include=FALSE}
rm(TCGA_RNA)
log_chunk_end_time("Clean-up TGCGA RNA", log_file = "chunk_times_log.txt")
```


# Disease selection and data normalization

```{r Disease_selection_and_data_normalization}

#------------ OLD CLASSIFICATION-------------------------------------------------

# selecionar apenas os dados da doença de interesse
#data <- selectDataPerDisease("TCGA-GBM", genes_expression, survival_data)
#data <- selectDataPerDisease("TCGA-LGG", genes_expression, survival_data)                                   

#------------ NEW CLASSIFICATION-------------------------------------------------

#data <- selectDataPerDisease("astrocytoma", genes_expression, survival_data)
#data <- selectDataPerDisease("glioblastoma", genes_expression, survival_data)
#data <- selectDataPerDisease("oligodendroglioma", genes_expression, survival_data)

#------------ CGGA CLASSIFICATION-------------------------------------------------

# keep only "IDH_mutation_status" = Wildtype, because to update GBM patients from the WHO-2016 to WHO-2021 classification the GBM patients need to be Wildtype
survival_data <- survival_data[survival_data$IDH_mutation_status == "Wildtype",]
survival_data <- survival_data[, -ncol(survival_data)] # apagar a coluna do "IDH_mutation_status" pq já não é necessária

data <- selectDataPerDisease(c("GBM", "rGBM"), genes_expression, survival_data)

genes_expression <- data$genes_expression
survival_data <- data$survival_data

# remove genes where the sum of the gene expression is 0 (necessary to repeat this step, since there is a selection of patients) 
genes_expression <- cleanGeneExpressionData(genes_expression)

# EdgeR+voom normalization
genes_expression <- normalizationEdgeR(genes_expression)

# Clean the gene expression dataframe
genes_expression <- cleanGeneExpressionData(genes_expression)

# ordenar os dados por paciente
genes_expression <- orderDataByPatient(genes_expression)
survival_data <- orderDataByPatient(survival_data)

# write on the chunk end time file
log_chunk_end_time("Disease selection and data normalization", log_file = "chunk_times_log.txt")
```

# Fit a Cox model with the desired genes (train =! test)

```{r Fit_a_Cox_model_with_the_desired_genes, message=FALSE, warning=FALSE}

#RO60 = TROVE2

# Get the coefficients of the best model
best_coefficients <- c("LOXL1", "PCDHB3", "TROVE2", "TNFRSF18", 
                       "IGFBP6", "NOSTRIN", "CADM3", "CEP97", "MMP19") 

# Split the data
splited <- splitTestAndTrain(genes_expression, survival_data, 0.7, seed_split = 6) # 3 ou 6 <- colocar o seed

genes_expression_train <- splited$expression_train
survival_train <- splited$survival_train
genes_expression_test <- splited$expression_test
survival_test <- splited$survival_test

# Define survival object
survival_object_train <- Surv(time = survival_train$days, event = survival_train$vital_status)

# Fit a Cox regression model using the covariates
fit <- coxph(survival_object_train ~ .,
             data = subset(genes_expression_train, select = best_coefficients)) # specify coefficients

# Test the proportional hazards assumption for a Cox regression model fit (coxph)
#Propo_hazards <- cox.zph(fit, transform="km", terms=TRUE, singledf=FALSE, global=TRUE)
#Propo_hazards_global_p_value <- Propo_hazards$table["GLOBAL", "p"]
#print(Propo_hazards$table)

# Construct a risk score based on the linear predictor on the test data
survival_probabilities_test <- predict(fit, newdata = subset(genes_expression_test, select = best_coefficients), type = "lp")
survival_probabilities_train <- predict(fit, newdata = subset(genes_expression_train, select = best_coefficients), type = "lp")


# Visualize the distribution of the risk predictor
plot_distribution <- ggplot(data.frame(Risk_Predictor = survival_probabilities_test), aes(x = Risk_Predictor)) +
                            geom_histogram(binwidth = 0.1, fill = "skyblue", color = "black") +
                            geom_density(color = "red", adjust = 1.5) +
                            labs(title = "Distribution of Risk Predictor",
                                 x = "Risk Predictor (Linear Predictor)",
                                 y = "Frequency") +
                            theme_minimal()

plot_distribution

# Categorize individuals of the test data based on the median
risk <- ifelse(survival_probabilities_test > median(survival_probabilities_train), "High", "Low")

# Kaplan-Meier com a separação por High/ Low
fit_surv <- survfit(Surv(survival_test$days/365.25, survival_test$vital_status) ~ risk)
survdiff(Surv(survival_test$days/365.25, survival_test$vital_status) ~ risk)

survdiff_output <- capture.output(survdiff(Surv(survival_test$days/365.25, survival_test$vital_status) ~ risk))
plot <- ggsurvplot(fit_surv,
                   data = survival_test,
                   title = "Kaplan-Meier Curves by Risk Status",
                   xlab = "Time (years)",
                   legend.title = "Group",
                   pval = TRUE)
plot$plot

ggsave(filename = "CGGA_Results/train_differ_test_plot_high_low_risk.pdf", plot = plot$plot, width = 6, height = 4)

```
# Fit a Cox model with the desired genes (train == test)

```{r Fit_a_Cox_model_with_the_desired_genes, message=FALSE, warning=FALSE}

#RO60 = TROVE2

# Get the coefficients of the best model
best_coefficients <- c("LOXL1", "PCDHB3", "TROVE2", "TNFRSF18", 
                       "IGFBP6", "NOSTRIN", "CADM3", "CEP97", "MMP19") 

# Define survival object
survival_object <- Surv(time = survival_data$days, event = survival_data$vital_status)

# Fit a Cox regression model using the covariates
fit <- coxph(survival_object ~ .,
             data = subset(genes_expression, select = best_coefficients)) # specify coefficients

# Test the proportional hazards assumption for a Cox regression model fit (coxph)
#Propo_hazards <- cox.zph(fit, transform="km", terms=TRUE, singledf=FALSE, global=TRUE)
#Propo_hazards_global_p_value <- Propo_hazards$table["GLOBAL", "p"]
#print(Propo_hazards$table)

# Construct a risk score based on the linear predictor on the test data
survival_probabilities <- predict(fit, newdata = subset(genes_expression, select = best_coefficients), type = "lp")

# Visualize the distribution of the risk predictor
plot_distribution <- ggplot(data.frame(Risk_Predictor = survival_probabilities), aes(x = Risk_Predictor)) +
                            geom_histogram(binwidth = 0.1, fill = "skyblue", color = "black") +
                            geom_density(color = "red", adjust = 1.5) +
                            labs(title = "Distribution of Risk Predictor",
                                 x = "Risk Predictor (Linear Predictor)",
                                 y = "Frequency") +
                            theme_minimal()

plot_distribution

# Categorize individuals of the test data based on the median
risk <- ifelse(survival_probabilities > median(survival_probabilities), "High", "Low")

# Kaplan-Meier com a separação por High/ Low
fit_surv <- survfit(Surv(survival_data$days/365.25, survival_data$vital_status) ~ risk)
survdiff(Surv(survival_data$days/365.25, survival_data$vital_status) ~ risk)

survdiff_output <- capture.output(survdiff(Surv(survival_data$days/365.25, survival_data$vital_status) ~ risk))
plot <- ggsurvplot(fit_surv,
                   data = survival_data,
                   title = "Kaplan-Meier Curves by Risk Status",
                   xlab = "Time (years)",
                   legend.title = "Group",
                   pval = TRUE)
plot$plot

ggsave(filename = "CGGA_Results/train_equal_test_plot_high_low_risk.pdf", plot = plot$plot, width = 6, height = 4)

```
